Panic on Type Assertion with Interface

## Description:
**Issue Summary:**

A panic occurs when performing a type assertion on an `interface{}` in Gnolang. The following code works as expected in Go (Golang), but in Gnolang, it results in a runtime panic:

```go
package main

import "fmt"

func main() {
	var i interface{} = 1
	var a, ok = i.(int)
	fmt.Println(a, ok)
}
```

**Expected Behavior:**

The code should execute without panic, correctly performing the type assertion. The expected output should be:

```
1 true
```

**Actual Behavior:**

In Gnolang, running the above code causes a panic with the message:

```
panic: interface conversion: gnolang.Expr is *gnolang.TypeAssertExpr, not *gnolang.CallExpr
```

This suggests an issue with how type assertions are being handled internally, specifically when converting types within the interface.

--- a/gnovm/pkg/gnolang/preprocess.go
+++ b/gnovm/pkg/gnolang/preprocess.go
@@ -2163,12 +2163,33 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				numNames := len(n.NameExprs)
 				sts := make([]Type, numNames) // static types
 				tvs := make([]TypedValue, numNames)
+
 				if numNames > 1 && len(n.Values) == 1 {
-					// special case if `var a, b, c T? = f()` form.
-					cx := n.Values[0].(*CallExpr)
-					tt := evalStaticTypeOfRaw(store, last, cx).(*tupleType)
-					if rLen := len(tt.Elts); rLen != numNames {
-						panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %s returns %d value(s)", numNames, cx.Func.String(), rLen))
+					// Special cases if one of the following:
+					// - `var a, b, c T = f()`
+					// - `var a, b = n.(T)`
+					// - `var a, b = n[i], where n is a map`
+
+					var tuple *tupleType
+					valueExpr := n.Values[0]
+					switch expr := valueExpr.(type) {
+					case *CallExpr:
+						tuple = evalStaticTypeOfRaw(store, last, expr).(*tupleType)
+					case *TypeAssertExpr, *IndexExpr:
+						valueType := evalStaticTypeOfRaw(store, last, expr)
+						tuple = &tupleType{Elts: []Type{valueType, BoolType}}
+					default:
+						panic(fmt.Sprintf("unexpected ValueDecl value expression type %T", expr))
+
+					}
+
+					if rLen := len(tuple.Elts); rLen != numNames {
+						rhsExprString := valueExpr.String()
+						if cx, ok := valueExpr.(*CallExpr); ok {
+							rhsExprString = cx.Func.String()
+						}
+
+						panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %s returns %d value(s)", numNames, rhsExprString, rLen))
 					}
 					if n.Type != nil {
 						// only a single type can be specified.
@@ -2181,7 +2202,7 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 					} else {
 						// set types as return types.
 						for i := 0; i < numNames; i++ {
-							et := tt.Elts[i]
+							et := tuple.Elts[i]
 							sts[i] = et
 							tvs[i] = anyValue(et)
 						}

--- a/gnovm/pkg/gnolang/preprocess.go
+++ b/gnovm/pkg/gnolang/preprocess.go
@@ -2189,13 +2189,16 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 					}
 
 					if rLen := len(tuple.Elts); rLen != numNames {
-						rhsExprString := valueExpr.String()
-						if cx, ok := valueExpr.(*CallExpr); ok {
-							rhsExprString = cx.Func.String()
-						}
-
-						panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %s returns %d value(s)", numNames, rhsExprString, rLen))
+						panic(
+							fmt.Sprintf(
+								"assignment mismatch: %d variable(s) but %s returns %d value(s)",
+								numNames,
+								valueExpr.String(),
+								rLen,
+							),
+						)
 					}
+
 					if n.Type != nil {
 						// only a single type can be specified.
 						nt := evalStaticType(store, last, n.Type)

--- a/gnovm/pkg/gnolang/preprocess.go
+++ b/gnovm/pkg/gnolang/preprocess.go
@@ -2172,15 +2172,20 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 
 					var tuple *tupleType
 					valueExpr := n.Values[0]
+					valueType := evalStaticTypeOfRaw(store, last, valueExpr)
+
 					switch expr := valueExpr.(type) {
 					case *CallExpr:
-						tuple = evalStaticTypeOfRaw(store, last, expr).(*tupleType)
+						tuple = valueType.(*tupleType)
 					case *TypeAssertExpr, *IndexExpr:
-						valueType := evalStaticTypeOfRaw(store, last, expr)
 						tuple = &tupleType{Elts: []Type{valueType, BoolType}}
+						if ex, ok := expr.(*TypeAssertExpr); ok {
+							ex.HasOK = true
+							break
+						}
+						expr.(*IndexExpr).HasOK = true
 					default:
 						panic(fmt.Sprintf("unexpected ValueDecl value expression type %T", expr))
-
 					}
 
 					if rLen := len(tuple.Elts); rLen != numNames {

--- a/gnovm/tests/files/var20.gno
+++ b/gnovm/tests/files/var20.gno
@@ -9,4 +9,4 @@ func main() {
 }
 
 // Error:
-// main/files/var20.gno:8:6: assignment mismatch: 3 variable(s) but r<VPBlock(3,0)> returns 1 value(s)
+// main/files/var20.gno:8:6: assignment mismatch: 3 variable(s) but r<VPBlock(3,0)>() returns 1 value(s)

--- a/gnovm/tests/files/vardecl.gno
+++ b/gnovm/tests/files/vardecl.gno
@@ -1,0 +1,23 @@
+package main
+
+func main() {
+	var i interface{} = 1
+	a, ok := i.(int)
+	println(a, ok)
+
+	b, c := doSomething()
+	println(b, c)
+
+	m := map[string]int{"a": 1, "b": 2}
+	d, okk := m["d"]
+	println(d, okk)
+}
+
+func doSomething() (int, string) {
+	return 4, "hi"
+}
+
+// Output:
+// 1 true
+// 4 hi
+// 0 false

--- a/gnovm/tests/files/vardecl.gno
+++ b/gnovm/tests/files/vardecl.gno
@@ -2,14 +2,14 @@ package main
 
 func main() {
 	var i interface{} = 1
-	a, ok := i.(int)
+	var a, ok = i.(int)
 	println(a, ok)
 
-	b, c := doSomething()
+	var b, c = doSomething()
 	println(b, c)
 
 	m := map[string]int{"a": 1, "b": 2}
-	d, okk := m["d"]
+	var d, okk = m["d"]
 	println(d, okk)
 }
 

